File: heap/295. Find Median from Data Stream.py
class MedianFinder:

    def __init__(self):
        self.arr = []

    def lower_bound(self, target):
        left, right = 0, len(self.arr) - 1
        while left <= right:
            mid = (left + right) // 2
            if self.arr[mid] < target:
                left = mid + 1
            else:
                right = mid - 1
        return left
    
    def addNum(self, num: int) -> None:
        self.arr.insert(self.lower_bound(num), num)

    def findMedian(self) -> float:
        n = len(self.arr)
        if n % 2 == 0:
            return (self.arr[n // 2] + self.arr[n // 2 - 1]) / 2
        return self.arr[n // 2]

File: heap/295. Find Median from Data Stream.java
import java.util.PriorityQueue;

class MedianFinder {
    private PriorityQueue<Integer> leftHalf;
    private PriorityQueue<Integer> rightHalf;

    public MedianFinder() {
        leftHalf = new PriorityQueue<>((a, b) -> b - a); // maxHeap
        rightHalf = new PriorityQueue<>(); // minHeap (default)
    }

    public void addNum(int num) {
        if (leftHalf.isEmpty() || num < leftHalf.peek()) {
            leftHalf.add(num);
        } else {
            rightHalf.add(num);
        }

        if (leftHalf.size() - rightHalf.size() > 1) {
            rightHalf.add(leftHalf.poll());
        } else if (rightHalf.size() - leftHalf.size() > 1) {
            leftHalf.add(rightHalf.poll());
        }
    }

    public double findMedian() {
        if (leftHalf.size() == rightHalf.size()) {
            return (leftHalf.peek() + rightHalf.peek()) / 2.0;
        }
        return leftHalf.size() > rightHalf.size() ? leftHalf.peek() : rightHalf.peek();
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */